### PR TITLE
Add option to not follow symlinks

### PR DESCRIPTION
**Pull request by kellyselden**
I believe symlinks used to be preserved, but now are followed. This is hurting me because in broccolijs land, everything is a symlink. This makes using the node-resolve plugin a nightmare.

Can we discuss an opt-in option for preserving symlinks? This PR is just a sample of what would change, not a final product. I would like the option to by system-wide public API, because several rollup plugins also follow rollup's example and follow symlinks. It would be nice if they could inherit the public option.

### CODE REVIEW

**Comment by kellyselden** [comment]
I added the opt-in option and added a test. It is also based on #1820 so all plugins can also read this option.

**Comment by kzc** [comment]
Is there a CLI flag for this setting without needing a config file? I can't get it to work.

**Comment by kellyselden** [comment]
@kzc I forgot about the CLI when I wrote this. It is just a config.

**Comment by kzc** [comment]
In light of [this description](https://github.com/rollup/rollup/issues/1555#issuecomment-322862209) it's surprising that Rollup has the node module resolution symlink behavior by default. It seems to be a bug. This PR should not be needed if Rollup behaved as described.

The node `realpathSync` stuff should only be in `rollup-plugin-node-resolve`, not in core `rollup`.

**Comment by guybedford** [comment]
Agreed this sounds like a bug to me. The default behaviour should ideally match Node resolution which would not be to use realpath, and rather have that be through a flag or plugin.

**Comment by kzc** [comment]
@Rich-Harris Was the use of node's `realpath` module resolution in Rollup core intentional?

https://github.com/rollup/rollup/blob/ddb698c94d74412bd5b788c860e24b2fe5639f12/src/utils/defaults.ts#L13

**Comment by Rich-Harris** [comment]
I can't recall to be honest, perhaps I was dealing with some tricky `npm link` scenario and that seemed to fix it... if the consensus is that we shouldn't use it then I'm on board 👍

**Comment by guybedford** [comment]
Perhaps we can consider then making `preserveSymlinks` the default behaviour to match Node.

This option could then either be `preserveSymlinks: false` to opt-out, or the inverse - `realpath: true`.

**Comment by kellyselden** [comment]
@guybedford Would that be an acceptable breaking change?

**Comment by kzc** [comment]
> Perhaps we can consider then making preserveSymlinks the default behaviour to match Node.

I find the phrase "to match node" to be a bit ambiguous. At the risk of restating what you wrote, I suggest by default Rollup should behave like `node --preserve-symlinks`. The use of `realpath` in Rollup should be an explicit opt in.

(Off topic: I wish that node flag had been named `--ignore-symlinks` as its intent was to treat symlinked files and directories as regular files and directories rather than invoking the confusion of "preserving" legacy node `realpath` module resolution.)

**Comment by guybedford** [comment]
Thanks @kzc for the clarifications and sorry for the confusion here, I got this the wrong way around then.

In that case, this PR exactly as it stands sounds like exactly the right approach to me, as the Node behaviour should be the default followed behaviour.

**Comment by kzc** [comment]
> In that case, this PR exactly as it stands sounds like exactly the right approach to me, as the Node behaviour should be the default followed behaviour.

Respectfully, I beg to differ. I think by default Rollup should not use "realpath" as this PR does. "realpath" should be an explicit opt-in.

Regardless of the default chosen, a Rollup CLI flag should be added to enable or disable "realpath" resolution.

**Comment by guybedford** [comment]
@kzc as far as I can tell, and from your clarification, this PR provides a `preserveSymlinks` just like NodeJS, which would default to `false`, just like Node.

Can you perhaps explain your reasoning a little on this? The reason NodeJS uses symlinks is because when linking one project into another, you still want the dependency resolution from the root project for peer dependencies, not end up loading duplicate dependencies between linked projects. So it's a core part of the integrity of the node_modules version management contract. Rollup should abide by that in my opinion, but have the opt-out available through this flag.

**Comment by guybedford** [comment]
Ugh, I will make sure to think before typing here! Comment removed for inaccuracies until I can actually take the time to put together a coherent thought :P

**Comment by kzc** [comment]
@guybedford No worries. The entire topic of "preserve symlinks" and node "realpath" module resolution is very confusing in general.

**Comment by guybedford** [comment]
So my the last deleted comment was on track with the arguments, I just got the wrong interpretation of symlink handling in NodeJS.

Basically, NodeJS follows symlinks to their real paths because when you link packages in a development workflow, you need to get the guarantee that the linked package dependencies will be used, so that multi-version between linked projects works.

Rollup should follow NodeJS here so that this same invariant is maintained when running Rollup in linking workflows this way.

This PR implements `preserveSymlinks: true` as an option to not use realpath, which then exactly matches NodeJS.

So my recommendation is to merge this as-is.

Apologies for confusions :)

**Comment by kzc** [comment]
> This PR implements `preserveSymlinks: true` as an option to not use `realpath`, which then exactly matches NodeJS.

@guybedford Just to confirm, did you mean the following?

"This PR implements `preserveSymlinks: true` as an option to not use `realpath`, which then exactly matches the behavior of `node --preserve-symlinks`"

Node in its default mode uses "realpath" module resolution (not to be confused with absolute path). If `node --preserve-symlinks` is run then node will **not** use "realpath" module resolution and treat symlinked files as if they were regular files and directories and resolve to absolute paths. This PR in its present state would have Rollup continue to use "realpath" module resolution by default.

Although I happen to disagree with the default chosen by this PR to retain "realpath" module resolution by default, I would still recommend this PR implement a command line flag to override this "realpath" behavior globally rather than having it solely settable via rollup config.

**Comment by kellyselden** [comment]
> Although I happen to disagree with the default chosen by this PR to retain "realpath" module resolution by default, I would still recommend this PR implement a command line flag to override this "realpath" behavior globally rather than having it solely settable via rollup config.

@kzc I'll do that once consensus is made on whether it is opt-in or opt-out.

**Comment by kzc** [comment]
Whatever @guybedford decides for default behavior is fine with me. No doubt some Rollup projects are relying on the default "realpath" module resolution behavior.

File: src/utils/first.ts [hunk 1]
```diff
@@ -1,11 +1,11 @@
 // Return the first non-null or -undefined result from an array of
 // maybe-sync, maybe-promise-returning functions
 export default function first<T> (candidates: ((...args: any[]) => Promise<T | void> | T | void)[]): (...args: any[]) => Promise<T | void> {
-	return function (...args: any[]) {
+	return function (this: any, ...args: any[]) {
 		return candidates.reduce((promise, candidate) => {
 			return promise.then(
 				result =>
-					result != null ? result : Promise.resolve(candidate(...args))
+					result != null ? result : Promise.resolve(candidate.apply(this, args))
```
Comment (guybedford) [inline src/utils/first.ts]: Are these changes intended?

File: src/utils/mergeOptions.ts [hunk 2]
```diff
@@ -74,6 +74,7 @@ export default function mergeOptions ({
 		cache: getInputOption('cache'),
 		preferConst: getInputOption('preferConst'),
 		experimentalDynamicImport: getInputOption('experimentalDynamicImport'),
+		preserveSymlinks: config.preserveSymlinks,
```
Comment (guybedford) [inline src/utils/mergeOptions.ts]: `getInputOption` here?

Comment (guybedford) [reply src/utils/mergeOptions.ts]: (as far as I can tell this would enable `--preserveSymlinks` in CLI but I may be reading this wrong?)

Comment (kzc) [reply src/utils/mergeOptions.ts]: I wasn't able to get the CLI option working. Is there some special syntax required?

Comment (guybedford) [reply src/utils/mergeOptions.ts]: Did you build with `getInputOptions('preserveSymlinks')`?

Comment (kzc) [reply src/utils/mergeOptions.ts]: No. I thought you were referring to the code as it exists.

File: src/utils/first.ts (hunk 1 above)
Comment (kellyselden) [reply src/utils/first.ts]: It comes from #1820 (which should be merged first IMO). It is so plugins that also do realpath (multi-entry, node-resolve) can read the option via `this.preserveSymlinks` and act accordingly, instead of each having their own duplicate options.

File: src/utils/mergeOptions.ts (hunk 2 above)
Comment (kellyselden) [reply src/utils/mergeOptions.ts]: Changed to `getInputOption`.

File: src/utils/first.ts (hunk 1 above)
Comment (kellyselden) [reply src/utils/first.ts]: It's also needed for https://github.com/rollup/rollup/pull/1819/files#diff-9b4fdd2a9bd0304295b89fa6270add84R54 to have `this` context.

**Comment by kellyselden** [comment]
I will take out the dependency on the `first` context.

**Comment by kellyselden** [comment]
rebased, and used curried options to remove dependency on `this`. Let me know what you think @lukastaegert

**Comment by lukastaegert** [comment]
@kellyselden Just checked your test and it does not turn red if I change the "preserveSymlinks" option. Which is probably because git did not restore any symlinks for me. As [the answers here](https://stackoverflow.com/questions/954560/how-does-git-handle-symbolic-links) sound like this should actually work, could you please check on your side that the symlink is actually correctly checked in?

**Comment by kellyselden** [comment]
@lukastaegert Fixed the test. When I rebased in Windows, it must have destroyed the symlink...

**Comment by lukastaegert** [comment]
It's working now, thanks 👍
